controllers/utils/copy_trigger_annotations.go

```diff
@@ -0,0 +1,134 @@
+	return ok
+}
+
+func GetCopyTriggerValue(pvc *corev1.PersistentVolumeClaim) string {
+	return pvc.Annotations[CopyTriggerAnnotation]
+}
+
+func GetLatestCopyTriggerValue(pvc *corev1.PersistentVolumeClaim) string {
+	return pvc.Annotations[LatestCopyTriggerAnnotation]
+}
+
+func GetLatestCopyTriggerWaitingSinceValue(pvc *corev1.PersistentVolumeClaim) string {
+	return pvc.Annotations[LatestCopyTriggerWaitingSinceAnnotation]
+}
+
+// Returns true if the pvc was updated
+func SetLatestCopyTriggerWaitingSinceValueNow(pvc *corev1.PersistentVolumeClaim) bool {
+	return setAnnotation(pvc, LatestCopyTriggerWaitingSinceAnnotation, time.Now().UTC().Format(time.RFC3339))
+}
+
+// Returns true if the pvc was updated
+func SetLatestCopyStatusWaitingForTrigger(pvc *corev1.PersistentVolumeClaim) bool {
+	updated := setAnnotation(pvc, LatestCopyStatusAnnotation, LatestCopyStatusValueWaitingForTrigger)
+	if updated || GetLatestCopyTriggerWaitingSinceValue(pvc) == "" {
+		// We're setting the status to WaitingForTrigger - set a waiting-since annotation
+		// to give us a timestamp
+		updated = SetLatestCopyTriggerWaitingSinceValueNow(pvc)
+	}
+
+	return updated
+}
+
+// Returns true if the pvc was updated
+func SetLatestCopyStatusInProgress(pvc *corev1.PersistentVolumeClaim) bool {
+	updated := setAnnotation(pvc, LatestCopyStatusAnnotation, LatestCopyStatusValueInProgress)
+	// Make sure waiting since annotation is removed
+	return unsetAnnotation(pvc, LatestCopyTriggerWaitingSinceAnnotation) || updated
+}
+
+// Returns true if the pvc was updated
+func SetLatestCopyStatusCompleted(pvc *corev1.PersistentVolumeClaim) bool {
+	updated := setAnnotation(pvc, LatestCopyStatusAnnotation, LatestCopyStatusValueCompleted)
+	// Make sure waiting since annotation is removed
+	return unsetAnnotation(pvc, LatestCopyTriggerWaitingSinceAnnotation) || updated
+}
+
+// Returns true if the pvc was updated
+func SetLatestCopyTriggerValue(pvc *corev1.PersistentVolumeClaim, value string) bool {
+	return setAnnotation(pvc, LatestCopyTriggerAnnotation, value)
+}
+
+// Returns true if the local obj resource was updated - does not perform a client.Update()
+func setAnnotation(obj metav1.Object, annotationName string, annotationValue string) bool {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	existingAnnotationValue, ok := annotations[annotationName]
+	if ok && existingAnnotationValue == annotationValue {
+		return false // Nothing to update
+	}
+
+	// Set the annotation and update
+	annotations[annotationName] = annotationValue
+	obj.SetAnnotations(annotations)
+
+	return true // updated
+}
+
+func unsetAnnotation(obj metav1.Object, annotationName string) bool {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+
+	_, ok := annotations[annotationName]
+	if !ok {
+		// Nothing to remove
+		return false
+	}
+
+	delete(annotations, annotationName)
```

Comment:
I'm trying to reconcile this w/ the setAnnotation()... in set, you call obj.SetAnnotations(), but not here. Is that because the SetAnnotations is only actually required in the case where the map was nil?

Reply:
I didn't intentionally do it differently, I think in unset I figured I'm updating the map directly - let me take a look, maybe I should do SetAnnotations in both places to be consistent.